dotcom-rendering/src/web/components/SlotBodyEnd.importable.tsx

```diff
@@ -88,12 +93,15 @@ export const SlotBodyEnd = ({
 	isPaidContent,
 	tags,
 	contributionsServiceUrl,
-	brazeMessages,
 	idApiUrl,
 	stage,
 	pageId,
 	keywordsId,
-}: Props) => {
+}: Omit<Props, 'isDev' | 'switches' | 'isSensitive'>) => {
+	const brazeMessages = useSWRPromise('braze-messages', () =>
+		buildBrazeMessages(idApiUrl),
```

Comment:
To make sure I understand correctly - does the use of `useSWRPromise` here effectively memoize the async work which `buildBrazeMessages` does? Under the hood an instance of `BrazeMessages` is returned which wraps the Braze SDK and I think we only ever want to have one of these on the page at a time. This used to happen because the instance was created in `App` and was passed down to the relevant "slots" as a prop.

I think a difference I can see with the new version of this code is that we'll only call `buildBrazeMessages` if one of (or both of) `SlotBodyEnd` and `StickyBottomBanner` exist in the page hierarchy? That's slightly different from before in `App` where we would _always_ call `buildBrazeMessages`. The reason that distinction is important is because even if the slots/islands don't exist on the page, we still want to call `buildBrazeMessage` currently as it does work to clear up after the SDK if the user has logged out or removed permissions for Braze to load. Any thoughts on how we could maintain that with this new approach? It's maybe not _super_ relevant at the moment because `StickBottomBanner` probably exists in every page layout? Even so it'd be nice to settle on an approach which works if that ever isn't true (now or in future).

Reply:
Yes, `useSWRPromise` memoize the async work of `buildBrazeMessages`. We might be able to use [`useSWRImmutable`][] under the hood to make sure there’s only ever one instance of the BrazeMEssages.

[`useSWRImmutable`]: https://swr.vercel.app/docs/revalidation#disable-automatic-revalidations

When it comes to making sure something is always called on a page, the current approach is having a script in the browser folder, like we do for `bootCmp` #3661, `initDiscussion` #3873 and [`coreVitals`][]

[`coreVitals`]: https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/web/browser/coreVitals/init.ts

Reply:
Worth noting that there's no standard way for code inside the Island to talk to code running through the script tags. If the fact that we have a Slot island on every page is not sufficient and we go the script way, implementing a mediator is a requirement.

One proposal for such a mediator is https://github.com/guardian/commercial-core/discussions/486.

Reply:
> The reason that distinction is important is because even if the slots/islands don't exist on the page, we still want to call `buildBrazeMessage` currently as it does work to clear up after the SDK if the user has logged out or removed permissions for Braze to load. Any thoughts on how we could maintain that with this new approach?

This is a really interesting and useful point!

So the thing that SWR brings to the party for us is request deduplication. It takes any promised based function along with a `key` value and wraps the call to it in some dedupe logic such that any subsequent call will return the same result.

The fact that this SWR takes any function, not just a fetch call, is powerful. It means we can dedupe kind of anything. This is potentially very powerful for our islands architecture. Where we have static global state shared between islands if we wrap the call to get this state using SWR then we effectively share it. Now, I used the words 'static' and 'effectively' there because the limitation to this SWR trick is we don't share dynamic changes to the state. If one island mutates it's state value the other islands are not aware of this.

To solve the issue you raised where we ideally always want to call this function, not just when a slot appears on the page, we could have a _third_ island, one that only makes the call to `buildBrazeMessages`. This will ensure that the call is always made but because of SWR deduplication it will not mean we make any additional actual executions of this function.

Reply:
> To solve the issue you raised where we ideally always want to call this function, not just when a slot appears on the page, we could have a third island, one that only makes the call to `buildBrazeMessages`

Ah cool, yeah that makes sense. If that fits with the way you (as in dotcom) are thinking about islands then I'd be in favour of adding that as part of this PR. That way even if it's true now that `StickBottomBanner` appears on ever page, it guards against that changing in future and going un-noticed. I wonder what a good name is? Would this be the first island where there's no visual/UI element to it?

Reply:
> Would this be the first island where there's no visual/UI element to it?

Not the first, no. [Liveness](https://github.com/guardian/dotcom-rendering/blob/4b1182429a41e65729f968cd10944b8595f773a3/dotcom-rendering/src/web/components/Liveness.importable.tsx) _sometimes_ returns a component but is mostly just a vehicle for code and [CommercialMetrics](https://github.com/guardian/dotcom-rendering/blob/8d0e112b92769bd6266312667c8c09377477c76c/dotcom-rendering/src/web/components/CommercialMetrics.importable.tsx) never returns any html. There might be others.